### PR TITLE
SONAR-8003 WS settings/set handles multi value settings

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/test/java/org/sonar/server/ws/ServletRequestTest.java [hunk 1]
```diff
@@ -19,57 +19,60 @@
...
+import static org.mockito.Mockito.when;
+
 public class ServletRequestTest {
 
   @Rule
   public ExpectedException expectedException = ExpectedException.none();
 
   HttpServletRequest source = mock(HttpServletRequest.class);
 
+  ServletRequest underTest = new ServletRequest(source);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/ws/ServletRequestTest.java]: Well done !

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/Request.java [hunk 2]
```diff
@@ -109,6 +109,15 @@ public long mandatoryParamAsLong(String key) {
     return values;
   }
 
+  public List<String> mandatoryMultiParam(String key) {
+    List<String> values = multiParam(key);
+    if (values.isEmpty()) {
+      throw new IllegalArgumentException(String.format("The '%s' parameter is missing", key));
```
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/Request.java]: Use Preconditions.checkArgument instead

Comment (teryk) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/Request.java]: should I in the API?

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java [hunk 3]
```diff
@@ -139,9 +151,33 @@ private String readParamOrDefaultValue(String key, @Nullable WebService.Param de
     return value;
   }
 
+  private List<String> readMultiParamOrDefaultValue(String key, @Nullable WebService.Param definition) {
+    if (definition == null) {
+      String message = String.format("BUG - parameter '%s' is undefined for action '%s'", key, action.key());
+      Loggers.get(getClass()).error(message);
```
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java]: I think this is an error to log this in log, it's not the responsability of this class to write errors in logs.

File: sonar-plugin-api/src/test/java/org/sonar/api/server/ws/RequestTest.java [hunk 4]
```diff
@@ -282,8 +303,9 @@ public void mandatory_param_as_part() throws Exception {
 
   private static class FakeRequest extends ValidatingRequest {
 
-    private final Map<String, String> params = Maps.newHashMap();
-    private final Map<String, Part> parts = Maps.newHashMap();
+    private final ListMultimap<String, String> multiParams = ArrayListMultimap.create();
```
Comment (julienlancelot) [inline sonar-plugin-api/src/test/java/org/sonar/api/server/ws/RequestTest.java]: Why didn't you do the same in TestRequest ?

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 5]
```diff
@@ -112,12 +169,19 @@ private void checkPermissions(Optional<ComponentDto> component) {
...
+    SetRequest setRequest = SetRequest.builder()
       .setKey(request.mandatoryParam(PARAM_KEY))
-      .setValue(request.mandatoryParam(PARAM_VALUE))
+      .setValue(request.param(PARAM_VALUE))
+      .setValues(request.multiParam(PARAM_VALUES))
       .setComponentId(request.param(PARAM_COMPONENT_ID))
       .setComponentKey(request.param(PARAM_COMPONENT_KEY))
       .build();
+
+    checkRequest((setRequest.getValue() != null && !setRequest.getValue().isEmpty())
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MAJOR NullPointerException might be thrown as 'getValue' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 6]
```diff
@@ -112,12 +169,19 @@ private void checkPermissions(Optional<ComponentDto> component) {
...
+      ^ (setRequest.getValues() != null && !setRequest.getValues().isEmpty()),
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MAJOR NullPointerException might be thrown as 'getValues' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 7]
```diff
@@ -130,10 +194,15 @@ private static SetRequest toWsRequest(Request request) {
     return Optional.of(project);
   }
 
-  private static PropertyDto toProperty(SetRequest request, Optional<ComponentDto> component) {
+  private PropertyDto toProperty(SetRequest request, Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 8]
```diff
@@ -120,19 +131,35 @@ private void validate(SetRequest request, Optional<ComponentDto> component) {
...
-    checkRequest(result.isValid(),
-      i18n.message(Locale.ENGLISH, "property.error." + result.getErrorKey(), "Error when validating setting with key '%s' and value '%s'"),
+    List<String> values = valuesFromRequest(request);
+    Optional<PropertyDefinition.Result> failingResult = values.stream()
+      .map(definition::validate)
+      .filter(result -> !result.isValid())
+      .findAny();
+    String errorKey = failingResult.isPresent() ? failingResult.get().getErrorKey() : null;
+
+    checkRequest(definition.multiValues() ^ request.getValue() != null,
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: Please move checks in another methods as it becomes less and less readable

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 9]
```diff
@@ -120,19 +131,35 @@ private void validate(SetRequest request, Optional<ComponentDto> component) {
...
   }
 
+  private static List<String> valuesFromRequest(SetRequest request) {
+    return request.getValue() == null ? request.getValues() : Collections.singletonList(request.getValue());
+  }
+
+  private static String persistedValue(List<String> values) {
+    return values.size() == 1
+      ? values.get(0)
+      : COMMA_JOINER.join(values.stream().map(value -> value.replace(",", "%2C")).collect(Collectors.toList()));
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: The replacement of %2C doesn't work when there's only one value

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 10]
```diff
@@ -120,19 +131,35 @@ private void validate(SetRequest request, Optional<ComponentDto> component) {
...
       "Setting '%s' cannot be set on a component with qualifier '%s'", request.getKey(), qualifier);
   }
 
+  private static List<String> valuesFromRequest(SetRequest request) {
+    return request.getValue() == null ? request.getValues() : Collections.singletonList(request.getValue());
+  }
+
+  private static String persistedValue(List<String> values) {
+    return values.size() == 1
+      ? values.get(0)
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: You forget to replace %2C by ","

Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: Hopefully my manual tests has detected this :)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java (hunk 3 above)
Comment (teryk) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java]: I did as done in other similar readParam. The behavior should be consistent, and I didn't have idea of impacts

File: sonar-plugin-api/src/test/java/org/sonar/api/server/ws/RequestTest.java (hunk 4 above)
Comment (teryk) [inline sonar-plugin-api/src/test/java/org/sonar/api/server/ws/RequestTest.java]: I did in the next commit, when using it

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java (hunk 10 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: argh! very good catch

File: server/sonar-server/src/test/java/org/sonar/server/ws/TestRequest.java [hunk 11]
```diff
@@ -48,8 +50,7 @@ protected String readParam(String key) {
 
   @Override
   protected List<String> readMultiParam(String key) {
-    String value = params.get(key);
-    return value == null ? emptyList() : singletonList(value);
+    return multiParams.get(key);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/ws/TestRequest.java]: 👍

File: sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SetRequest.java [hunk 12]
```diff
@@ -42,11 +45,17 @@ public String getKey() {
     return key;
   }
 
+  @CheckForNull
   public String getValue() {
     return value;
   }
 
   @CheckForNull
```
Comment (julienlancelot) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SetRequest.java]: This CheckForNull should be removed, no ?

File: sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SetRequest.java [hunk 13]
```diff
@@ -75,11 +85,16 @@ public Builder setKey(String key) {
       return this;
     }
 
-    public Builder setValue(String value) {
+    public Builder setValue(@Nullable String value) {
       this.value = value;
       return this;
     }
 
+    public Builder setValues(@Nullable List<String> values) {
```
Comment (julienlancelot) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SetRequest.java]: Why Nullable ? We should not accept null value

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/Request.java (hunk 2 above)
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/Request.java]: Ahhhhh, indeed, probably not in the API...

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java (hunk 3 above)
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java]: Then I tell you => we should not do that (See last email from Eric about failure with protobuff)

**Comment by julienlancelot** [comment]
Please apply my feedbacks then LGTM

File: sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SetRequest.java (hunk 13 above)
Comment (teryk) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SetRequest.java]: as we accept null value in the output, the `SetRequest.getValues()` can return null

File: sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SetRequest.java (hunk 12 above)
Comment (teryk) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SetRequest.java]: I think not, except if we want to return an empty list and checks it fails if we add an empty list

**Pull request merged**